<a href="https://colab.research.google.com/github/calarconf/AlgorithmsUN2024I/blob/main/CormenSizeTimeComplexityCalarconf.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Id: 1 \\
1 mod 4 = 1, por lo tanto hice el numeral 1 \\
Cristhian Alejandro Alarcón Florido

# 0.(1 microsecond by instruction)
Asuuming that you can execute a million instructions $10^6$ every second, i.e. each instruction takes one microsecond $1\mu s=10^{-6}$ seconds. If you have algorithms take $\bf{ Time\ Complexity}$ steps to solve a problem of size $n$ as specified in column one. What is the size of the largest problem that you can solve in the time specified in row one?

$$
\begin{array}{|c|c|c|c|c|c|c|c|}
 \hline
 \bf{ Time\ Complexity} & Second  & Minute  & Hour & Day & Year & Century & Mileniun  \\
  \hline\\
\lg(n) &    &   &  &  &  &  &  \\
   \hline\\
\lg^2(n) &    &   &  &  &  &  &  \\
   \hline\\
\sqrt{n} &    &   &  &  &  &  &  \\
 \hline
 n &    &   &  &  &  &  &  \\
 \hline
 n\lg(n) &    &   &  &  &  &  &  \\
 \hline
 n^2 &    &   &  &  &  &  &  \\
 \hline
  n^3 &    &   &  &  &  &  &  \\
 \hline
  n^4 &    &   &  &  &  &  &  \\
 \hline
  2^n &    &   &  &  &  &  &  \\
 \hline
 n! &    &   &  &  &  &  &  \\
 \hline
 n^n &    &   &  &  &  &  &  \\
 \hline
\end{array}
$$

# 1.(10  nanoseconds by instruction)

Asuuming that you can execute a 100 million instructions $100 \times 10^6$ every second, i.e. each instrcution takes 10 nanoseconds, 10 $\eta s  = 10^{-8} s $. If you have algorithms take $\bf{ Time\ Complexity}$ steps to solve a poblem of size $n$ as specified in column one. What is the size of the largest porblem that you can solve in the time specified in row one?
$$
\begin{array}{|c|c|c|c|c|c|c|c|}
 \hline
 \bf{ Time\ Complexity} & Second  & Minute  & Hour & Day & Year & Century & Mileniun  \\
  \hline\\
\lg(n) &    2^{1.00000000E+06} & 2^ {6.00000000E+07} & 2^{ 3.60000000E+09} & 2^ {8.64000000E+10} &  2^ {3.15360000E+1}  & 2^ {3.15360000E+15}  & 2^ {3.15360000E+18} \\
   \hline\\
\lg^2(n) & 2^{ 1000 }   &  2^{ 7746 }  & 2^{ 60000 } & 2^{ 293939 } & 2^{ 5615692 } & 2^{ 56156923 } & 2^{ 1775837830 } \\
   \hline\\
\sqrt{n} &  1.00000000E+12  &  3.60000000E+15 &  1.29600000E+19 &  7.46496000E+21 &  9.94519296E+26 & 9.94519296E+30 & 9.94519296E+36 \\
 \hline
 n & 1.00000000E+06   &  6.00000000E+07  & 3.60000000E+09 & 8.64000000E+10 & 3.15360000E+13 & 3.15360000E+15 & 3.15360000E+18 \\
 \hline
 n\lg(n) &  6.27460000E+04   & 2.80141700E+06  & 1.33378058E+08 & 2.75514751E+09 & 7.97633893E+11 &  6.86109568E+13 & 5.66650934E+16 \\
 \hline
 n^2 &  1000  & 7746 & 60000 & 293939 & 5615692 & 56156923 & 1775837830 \\
 \hline
  n^3 &  100  & 391 & 1533 & 4421 & 31594 & 146646 &  1466455 \\
 \hline
  n^4 &  32  & 88  & 245 & 542 &  2370 & 7494 & 42141 \\
 \hline
  2^n &   19 &  25 & 31 & 36 & 44 & 51 & 61 \\
 \hline
 n! &  9  & 11  & 12 & 13 & 16 & 17 & 20 \\
 \hline
 n^n &  7  & 8  & 9 & 10 & 12 & 13 & 15 \\
 \hline
\end{array}
$$

$\bf{ Time\ Complexity\ \lg(n) }$

$Second$
$$ \lg(n) \times 10^{-6}\ s\le 1\ s$$
$$ \lg(n) \le 1 \times 10^{6}$$
$$ 2^{\lg(n)} \le 2^{10^{6}}$$
$$ n \le 2^{10^{6}}$$

$Minute$
$$ \lg(n) \times 10^{-6} \ s \le 60\ s$$
$$ \lg(n) \le 60 \times 10^{6} $$
$$ 2^{\lg(n)} \le 2^{60 \times 10^{6} }$$
$$ n \le 2^{60 \times 10^{6} }$$

$Hour$
$$ \lg(n) \times 10^{-6} \ s \le 3600\ s$$
$$ \lg(n) \le 3600 \times 10^{6} $$
$$ 2^{\lg(n)} \le 2^{3.6 \times 10^{9}}$$
$$ n \le 2^{3.6 \times 10^{9}}$$

$Day$
$$ \lg(n) \times 10^{-6} \ s \le 86400\ s$$
$$ \lg(n) \le 86400\times 10^{6} $$
$$ 2^{\lg(n)} \le 2^{8.64 \times 10^{11}}$$
$$ n \le 2^{8.64 \times 10^{10}}$$

$Year$
$$ \lg(n) \times 10^{-6} \ s \le 31536000\ s$$
$$ \lg(n) \le 31536000\times 10^{6} $$
$$ 2^{\lg(n)} \le 2^{3.1536 \times 10^{14}}$$
$$ n \le 2^{3.1536 \times 10^{14}}$$

$Century$
$$ \lg(n) \times 10^{-6} \ s \le 3153600000\ s$$
$$ \lg(n) \le 3153600000\times 10^{6} $$
$$ 2^{\lg(n)} \le 2^{3.1536 \times 10^{16}}$$
$$ n \le 2^{3.1536 \times 10^{16}}$$

$Mileniun$
$$ \lg(n) \times 10^{-6} \ s \le 3153600000000\ s$$
$$ \lg(n) \le 3153600000000\times 10^{6} $$
$$ 2^{\lg(n)} \le 2^{3.1536 \times 10^{19}}$$
$$ n \le 2^{3.1536 \times 10^{19}}$$


In [1]:

print("Second  = 2^", "{:.8E}".format(10**6))
print("Minute  = 2^", "{:.8E}".format(60*10**6))
print("Hour  = 2^", "{:.8E}".format(60*60*10**6))
print("Day  = 2^", "{:.8E}".format(60*60*24*10**6))
print("Year  = 2^", "{:.8E}".format(365*24*60*60*10**6))
print("Century  = 2^", "{:.8E}".format(100*365*24*60*60*10**6))
print("Mileniun  = 2^", "{:.8E}".format(1000*100*365*24*60*60*10**6))

Second  = 2^ 1.00000000E+06
Minute  = 2^ 6.00000000E+07
Hour  = 2^ 3.60000000E+09
Day  = 2^ 8.64000000E+10
Year  = 2^ 3.15360000E+13
Century  = 2^ 3.15360000E+15
Mileniun  = 2^ 3.15360000E+18


$\bf{ Time\ Complexity\ \lg^2(n) }$

$Second$
$$ \lg^2(n) \times 10^{-6} \ s\le 1\ s$$
$$ \lg(n) \le \sqrt{1 \times 10^{6} }$$
$$ 2^{\lg(n)} \le 2^{\sqrt{1 \times 10^{6} }}$$
$$ n \le 2^{3162}$$

$Minute$
$$ \lg^2(n) \times 10^{-6} \ s \le 60\ s$$
$$ \lg(n) \le \sqrt{60 \times 10^{6} }$$
$$ 2^{\lg(n)} \le 2^{\sqrt{60 \times 10^{6} }}$$
$$ n \le 2^{ 24495}$$

$Hour$
$$ \lg^2(n) \times 10^{-6} \ s \le 3600\ s$$
$$ \lg(n) \le \sqrt{3600 \times 10^{6} }$$
$$ 2^{\lg(n)} \le 2^{\sqrt{3.6 \times 10^{9}}}$$
$$ n \le  2^{ 189737 }$$

$Day$
$$ \lg^2(n) \times 10^{-6} \ s \le 86400\ s$$
$$ \lg(n) \le \sqrt{86400\times 10^{6} }$$
$$ 2^{\lg(n)} \le 2^{\sqrt{8.6400\times 10^{10}}}$$
$$ n \le 2^{ 929516 }$$

$Year$
$$ \lg^2(n) \times 10^{-6} \ s \le 31536000\ s$$
$$ \lg(n) \le \sqrt{31536000\times 10^{6} }$$
$$ 2^{\lg(n)} \le 2^{\sqrt{3.1536\times 10^{13}}}$$
$$ n \le 2^{ 17758378 }$$

$Century$
$$ \lg^2(n) \times 10^{-6} \ s \le 3153600000\ s$$
$$ \lg(n) \le \sqrt{3153600000\times 10^{6} }$$
$$ 2^{\lg(n)} \le 2^{\sqrt{3.1536 \times 10^{15}}}$$
$$ n \le 2^{ 177583783 }$$

$Mileniun$
$$ \lg^2(n) \times 10^{-6} \ s \le 3153600000000\ s$$
$$ \lg(n) \le \sqrt{3153600000000\times 10^{6} }$$
$$ 2^{\lg(n)} \le 2^{\sqrt{3.1536 \times 10^{18}}}$$
$$ n \le 2^{ 5615692299 }$$

In [2]:
import math
print("Second 2^{",round(math.sqrt(10**6)),"}")
print("Minute 2^{",round(math.sqrt(60*10**6)),"}")
print("Hour 2^{",round(math.sqrt(60*60*10**6)),"}")
print("Day 2^{",round(math.sqrt(24*60*60*10**6)),"}")
print("Year 2^{",round(math.sqrt(365*24*60*60*10**6)),"}")
print("Century 2^{",round(math.sqrt(100*365*24*60*60*10**6)),"}")
print("Mileniun 2^{",round(math.sqrt(1000*100*365*24*60*60*10**6)),"}")


Second 2^{ 1000 }
Minute 2^{ 7746 }
Hour 2^{ 60000 }
Day 2^{ 293939 }
Year 2^{ 5615692 }
Century 2^{ 56156923 }
Mileniun 2^{ 1775837830 }


$\bf{ Time\ Complexity\ \sqrt{n} }$

$Second$
$$  \sqrt{n}  \times 10^{-6} \ s\le 1\ s$$
$$  \sqrt{n}  \le 10^{6}  $$
$$  n \le   10^{14}$$

$Minute$
$$  \sqrt{n} \times 10^{-6} \ s\le 60\ s$$
$$  \sqrt{n}  \le 60 \times 10^{6} $$
$$  n \le 3.600 \times10^{17}$$

$Hour$
$$  \sqrt{n}\times 10^{-6} \ s\le 3600\ s$$
$$  \sqrt{n} \le 3600 \times 10^{6}$$
$$  n  \le 1.296 \times10^{21}$$

$Day$
$$  \sqrt{n} \times 10^{-6} \ s\le 84600\ s$$
$$  \sqrt{n}\le 84600 \times 10^{6} $$
$$  n \le 7.15716 \times10^{23}$$

$Year$
$$  \sqrt{n} \times 10^{-6} \ s\le 31536000\ s$$
$$  \sqrt{n} \le 31536000 \times 10^{6} $$
$$  n \le 9.94519296 \times10^{28}$$

$Century$
$$  \sqrt{n} \times 10^{-6} \ s\le 3153600000\ s$$
$$  \sqrt{n} \le 3153600000 \times 10^{6} $$
$$  n \le 9.94519296  \times10^{32}$$

$Mileniun$
$$  \sqrt{n} \times 10^{-6} \ s\le 3153600000000\ s$$
$$  \sqrt{n} \le 3153600000000 \times 10^{6} $$
$$  n \le 9.94519296 \times10^{38}$$



In [3]:
import math
print("Second = ", "{:.8E}".format(math.pow(10**6,2)))
print("Minute = ","{:.8E}".format(math.pow(60*10**6,2)))
print("Hour = ","{:.8E}".format(math.pow(60*60*10**6,2)))
print("Day = ","{:.8E}".format(math.pow(24*60*60*10**6,2)))
print("Year = ","{:.8E}".format(math.pow(365*24*60*60*10**6,2)))
print("Century = ","{:.8E}".format(math.pow(100*365*24*60*60*10**6,2)))
print("Mileniun  = ","{:.8E}".format(math.pow(1000*100*365*24*60*60*10**6,2)))

Second =  1.00000000E+12
Minute =  3.60000000E+15
Hour =  1.29600000E+19
Day =  7.46496000E+21
Year =  9.94519296E+26
Century =  9.94519296E+30
Mileniun  =  9.94519296E+36


$\bf{ Time\ Complexity\ n }$

$Second$
$$  n  \times 10^{-6} \ s\le 1\ s$$
$$  n \le  10^{6} $$


$Minute$
$$  n \times 10^{-6} \ s\le 60\ s$$
$$  n \le 60 \times 10^{6} $$


$Hour$
$$  n \times 10^{-6} \ s\le 3600\ s$$
$$  n \le 3600 \times 10^{6}  = 3.6 \times 10^{10}$$


$Day$
$$  n \times 10^{-6} \ s\le 84600\ s$$
$$  n \le 84600 \times 10^{6}  = 8.46 \times 10^{11}$$


$Year$
$$  n \times 10^{-6} \ s\le 31536000\ s$$
$$  n \le 31536000 \times 10^{6}   = 3.153 \times 10^{14}$$



$Century$
$$  n \times 10^{-6} \ s\le 3153600000\ s$$
$$  n \le 3153600000 \times 10^{6}  = 3.153 \times 10^{16}$$


$Mileniun$
$$  n \times 10^{-6} \ s\le 3153600000000\ s$$
$$  n \le 3153600000000 \times 10^{6}  = 3.153 \times 10^{19}$$

In [5]:
import math
print("Second =", "{:.8E}".format(10**6))
print("Minute  =","{:.8E}".format(60*10**6))
print("Hour =","{:.8E}".format(60*60*10**6))
print("Day =","{:.8E}".format(24*60*60*10**6))
print("Year =","{:.8E}".format(365*24*60*60*10**6))
print("Century =","{:.8E}".format(100*365*24*60*60*10**6))
print("Mileniun =","{:.8E}".format(1000*100*365*24*60*60*10**6))

Second = 1.00000000E+06
Minute  = 6.00000000E+07
Hour = 3.60000000E+09
Day = 8.64000000E+10
Year = 3.15360000E+13
Century = 3.15360000E+15
Mileniun = 3.15360000E+18


$\bf{ Time\ Complexity\ n\lg{n} }$

$Second$
$$   n\lg{n}  \times 10^{-6} \ s\le 1\ s$$
$$   n\lg{n} \le  10^{6} $$


$Minute$
$$   n\lg{n} \times 10^{-6} \ s\le 60\ s$$
$$   n\lg{n} \le 60 \times 10^{6} $$


$Hour$
$$   n\lg{n} \times 10^{-6} \ s\le 3600\ s$$
$$   n\lg{n} \le 3600 \times 10^{6} $$


$Day$
$$   n\lg{n} \times 10^{-6} \ s\le 84600\ s$$
$$   n\lg{n} \le 84600 \times 10^{6} $$


$Year$
$$   n\lg{n} \times 10^{-6} \ s\le 31536000\ s$$
$$   n\lg{n} \le 31536000 \times 10^{6} $$



$Century$
$$   n\lg{n} \times 10^{-6} \ s\le 3153600000\ s$$
$$   n\lg{n} \le 3153600000 \times 10^{6} $$


$Mileniun$
$$   n\lg{n} \times 10^{-6} \ s\le 3153600000000\ s$$
$$  n \le 3153600000000 \times 10^{6} $$

In [7]:
import math
def f(n):
  return n*math.log2(n)


def find(f,v):
  l = 1
  r = v
  while True:
    m = int((l+r)/2)
    if ((f(m+1) > v) and (f(m) <= v) ) or (l == m) or  (r == m):
      return int(m)
    elif f(m) < v:
      l = m
    else:
      r = m

print("Second ", find(f,10**7)," = ", "{:.8E}".format(find(f,10**6 )))
print("Minute  ",find(f,60*10**7)," = ", "{:.8E}".format(find(f,60*10**6 )))
print("Hour ",find(f,60*60*10**7)," = ", "{:.8E}".format(find(f,60*60*10**6 )))
print("Day ",find(f,24*60*60*10**7 )," = ", "{:.8E}".format(find(f,24*60*60*10**6 )) )
print("Year ",find(f,365*24*60*60*10**7 )," = ", "{:.8E}".format(find(f,365*24*60*60*10**6 )) )
print("Century ",find(f,100*365*24*60*60*10**7 )," = ", "{:.8E}".format(find(f,100*365*24*60*60*10**6 )) )
print("Mileniun ",find(f,1000*100*365*24*60*60*10**7 )," = ", "{:.8E}".format(find(f,1000*100*365*24*60*60*10**6 )) )

Second = 526172  =  6.27460000E+04
Minute  = 24446750  =  2.80141700E+06
Hour = 1193909231  =  1.33378058E+08
Day = 25013054975  =  2.75514751E+09
Year = 7377486406427  =  7.97633893E+11
Century = 641136862352841  =  6.86109568E+13
Mileniun = 535474191787114112  =  5.66650934E+16


$\bf{ Time\ Complexity\ n^2}$

$Second$
$$  n^2 \times 10^{-6} \ s\le 1\ s$$
$$  n^2 \le  10^{6} $$
$$ \sqrt{n^2} \le \sqrt{10^{6} }$$
$$ n \le 1000 $$

$Minute$
$$  n^2\times 10^{-6} \ s\le 60\ s$$
$$  n^2 \le 60 \times 10^{6} $$
$$ \sqrt{n^2} \le \sqrt{60 \times10^{6} }$$
$$ n \le  7746 $$
  
$Hour$
$$  n^2 \times 10^{-6} \ s\le 3600\ s$$
$$ n^2 \le 3600 \times 10^{6} $$
$$ \sqrt{n^2} \le \sqrt{3.6 \times10^{9}}$$
$$ n \le 60000$$

$Day$
$$  n^2 \times 10^{-6} \ s\le 84600\ s$$
$$  n^2 \le 84600 \times 10^{6} $$
$$ \sqrt{n^2} \le \sqrt{8.46 \times10^{10}}$$
$$ n \le 293939 $$

$Year$
$$  n^2 \times 10^{-6} \ s\le 31536000\ s$$
$$  n^2 \le 31536000 \times 10^{6} $$
$$ \sqrt{n^2} \le \sqrt{3.1536 \times10^{13}}$$
$$ n \le 5615692 $$

$Century$
$$  n^2 \times 10^{-6} \ s\le 3153600000\ s$$
$$  n^2 \le 3153600000 \times 10^{6} $$
$$ \sqrt{n^2} \le \sqrt{3.1536 \times10^{15}}$$
$$ n \le 56156923 $$

$Mileniun$
$$  n^2 \times 10^{-6} \ s\le 3153600000000\ s$$
$$ n^2 \le 3153600000000 \times 10^{6} $$
$$ \sqrt{n^2} \le \sqrt{3.1536 \times10^{18}}$$
$$ n \le 1775837830 $$  


In [8]:
import math
print("Second =",round(math.sqrt(10**6 )))
print("Minute  =",round(math.sqrt(60*10**6 )))
print("Hour =",round(math.sqrt(60*60*10**6 )))
print("Day =",round(math.sqrt(24*60*60*10**6 )))
print("Year =",round(math.sqrt(365*24*60*60*10**6 )))
print("Century =",round(math.sqrt(100*365*24*60*60*10**6 )))
print("Mileniun =",round(math.sqrt(1000*100*365*24*60*60*10**6 )))

Second = 1000
Minute  = 7746
Hour = 60000
Day = 293939
Year = 5615692
Century = 56156923
Mileniun = 1775837830


$\bf{ Time\ Complexity\ n^3 }$

$Second$
$$  n^3 \times 10^{-6} \ s\le 1\ s$$
$$  n^3 \le  10^{6} $$
$$ \sqrt[4]{n^3} \le \sqrt[3]{10^{6} }$$
$$ n \le  100 $$

$Minute$
$$  n^3 \times 10^{-6} \ s\le 60\ s$$
$$  n^3 \le 60 \times 10^{6} $$
$$ \sqrt[3]{n^3} \le \sqrt[3]{60 \times10^{6} }$$
$$ n \le  391 $$
  
$Hour$
$$  n^3 \times 10^{-6} \ s\le 3600\ s$$
$$  n^3 \le 3600 \times 10^{6} $$
$$ \sqrt[3]{n^3 } \le \sqrt[3]{3.6 \times10^{9}}$$
$$ n \le 1532 $$

$Day$
$$  n^3 \times 10^{-6} \ s\le 84600\ s$$
$$  n^3 \le 84600 \times 10^{6} $$
$$ \sqrt[3]{n^3} \le \sqrt[3]{8.46 \times10^{10}}$$
$$ n \le 4421 $$

$Year$
$$  n^3 \times 10^{-6} \ s\le 31536000\ s$$
$$  n^3 \le 31536000 \times 10^{6} $$
$$ \sqrt[3]{n^3} \le \sqrt[3]{3.1536 \times10^{13}}$$
$$ n \le 31594 $$

$Century$
$$  n^3 \times 10^{-6} \ s\le 3153600000\ s$$
$$  n^3 \le 3153600000 \times 10^{6} $$
$$ \sqrt[3]{n^3} \le \sqrt[3]{3.1536 \times10^{15}}$$
$$ n \le 146646 $$

$Mileniun$
$$  n^3 \times 10^{-6} \ s\le 3153600000000\ s$$
$$  n^3 \le 3153600000000 \times 10^{6} $$
$$ \sqrt[3]{n^3} \le \sqrt[3]{3.1536 \times10^{18}}$$
$$ n \le 1466455 $$

In [10]:
import math
print("Second =",round(math.pow(10**6 ,1/3)))
print("Minute  =",round(math.pow(60*10**6 ,1/3)))
print("Hour =",round(math.pow(60*60*10**6 ,1/3)))
print("Day =",round(math.pow(24*60*60*10**6 ,1/3)))
print("Year =",round(math.pow(365*24*60*60*10**6 ,1/3)))
print("Century =",round(math.pow(100*365*24*60*60*10**6 ,1/3)))
print("Mileniun =",round(math.pow(1000*100*365*24*60*60*10**6 ,1/3)))



Second = 100
Minute  = 391
Hour = 1533
Day = 4421
Year = 31594
Century = 146646
Mileniun = 1466455


$\bf{ Time\ Complexity\ n^4 }$

$Second$
$$  n^4 \times 10^{-6} \ s\le 1\ s$$
$$  n^4\le  10^{6} $$
$$ \sqrt[4]{n^4 } \le \sqrt[4]{10^{6} }$$
$$ n \le   32 $$

$Minute$
$$  n^4 \times 10^{-6} \ s\le 60\ s$$
$$  n^4 \le 60 \times 10^{6} $$
$$ \sqrt[4]{n^4 } \le \sqrt[4]{60 \times10^{6} }$$
$$ n \le 88 $$
  
$Hour$
$$  n^4 \times 10^{-6} \ s\le 3600\ s$$
$$  n^4 \le 3600 \times 10^{6} $$
$$ \sqrt[4]{n^4} \le \sqrt[4]{3.6 \times10^{9}}$$
$$ n \le 245 $$  

$Day$
$$  n^4 \times 10^{-6} \ s\le 84600\ s$$
$$  n^4 \le 84600 \times 10^{6} $$
$$ \sqrt[4]{n^4} \le \sqrt[4]{8.46 \times10^{10}}$$
$$ n \le 542 $$

$Year$
$$  n^4 \times 10^{-6} \ s\le 31536000\ s$$
$$  n^4 \le 31536000 \times 10^{6} $$
$$ \sqrt[4]{n^4} \le \sqrt[4]{3.1536 \times10^{13}}$$
$$ n \le 2370 $$

$Century$
$$  n^4 \times 10^{-6} \ s\le 3153600000\ s$$
$$  n^4 \le 3153600000 \times 10^{6} $$
$$ \sqrt[4]{n^4} \le \sqrt[4]{3.1536 \times10^{15}}$$
$$ n \le 7494 $$

$Mileniun$
$$  n^4 \times 10^{-6} \ s\le 3153600000000\ s$$
$$  n^4 \le 3153600000000 \times 10^{6} $$
$$ \sqrt[4]{n^4} \le \sqrt[4]{3.1536 \times10^{18}}$$
$$ n \le 42141 $$

  

In [12]:
import math
print("Second =",round(math.pow(10**6 ,1/4)))
print("Minute  =",round(math.pow(60*10**6 ,1/4)))
print("Hour =",round(math.pow(60*60*10**6 ,1/4)))
print("Day =",round(math.pow(24*60*60*10**6 ,1/4)))
print("Year =",round(math.pow(365*24*60*60*10**6 ,1/4)))
print("Century =",round(math.pow(100*365*24*60*60*10**6 ,1/4)))
print("Mileniun =",round(math.pow(1000*100*365*24*60*60*10**6 ,1/4)))

Second = 32
Minute  = 88
Hour = 245
Day = 542
Year = 2370
Century = 7494
Mileniun = 42141


**texto en negrita**$\bf{ Time\ Complexity\ 2^n }$

$Second$
$$  2^n  \times 10^{-6} \ s\le 1\ s$$
$$  2^n \le  10^{6} $$
$$  \lg(2^n) \le  \lg(10^{6} )$$
$$  n \le  \lg(10^{6} ) = 19 $$

$Minute$
$$  2^n  \times 10^{-6} s \le 60\ s$$
$$  2^n \le   60 \times  10^{6} $$
$$  \lg(2^n) \le  \lg(60 \times  10^{6} )$$
$$  n \le  \lg(10^{60 \times  10^{6} }) = 25 $$


$Hour$
$$  2^n  \times 10^{-6} s \le 3600\ s$$
$$  2^n \le   3600 \times  10^{6} $$
$$  \lg(2^n) \le  \lg(3.6 \times  10^{9})$$
$$  n \le  \lg(10^{3.6 \times  10^{9}}) = 31 $$

$Day$
$$  2^n  \times 10^{-6} s \le  84600\ s$$
$$  2^n \le    84600\times  10^{6} $$
$$  \lg(2^n) \le  \lg( 8.46 \times  10^{10})$$
$$  n \le  \lg(10^{8.46 \times  10^{10}}) = 36 $$

$Year$
$$  2^n  \times 10^{-6} s \le  31536000\ s$$
$$  2^n \le    31536000\times  10^{6} $$
$$  \lg(2^n) \le  \lg( 3.1536 \times  10^{13})$$
$$  n \le  \lg(10^{3.1536 \times  10^{13}}) = 44 $$

$Century$
$$  2^n  \times 10^{-6} s \le  3153600000\ s$$
$$  2^n \le    3153600000\times  10^{6} $$
$$  \lg(2^n) \le  \lg( 3.1536 \times  10^{15})$$
$$  n \le  \lg(10^{3.1536 \times  10^{15}}) = 51 $$


$Mileniun$
$$  2^n  \times 10^{-6} s \le  3153600000000\ s$$
$$  2^n \le    3153600000000\times  10^{6} $$
$$  \lg(2^n) \le  \lg( 3.1536 \times  10^{18})$$
$$  n \le  \lg(10^{3.1536 \times  10^{18}}) = 61 $$

In [14]:
import math
print("Second =",int(math.log2(10**6 )))
print("Minute  =",int(math.log2(60*10**6 )))
print("Hour =",int(math.log2(60*60*10**6 )))
print("Day =",int(math.log2(24*60*60*10**6 )))
print("Year =",int(math.log2(365*24*60*60*10**6 )))
print("Century =",int(math.log2(100*365*24*60*60*10**6 )))
print("Mileniun =",int(math.log2(1000*100*365*24*60*60*10**6 )))

Second = 19
Minute  = 25
Hour = 31
Day = 36
Year = 44
Century = 51
Mileniun = 61


In [15]:
import math
def f(n):
  return 2**n

def find(f,v):
  l = 1
  r = 100
  while True:
    m = int((l+r)/2)
    if ((f(m+1) > v) and (f(m) <= v) ) or (l == m)  or (r== m):
      return int(m)
    elif f(m) < v:
      l = m
    else:
      r = m


print("Second ", find(f,10**6 ))
print("Minute  ",find(f,60*10**6 ))
print("Hour ",find(f,60*60*10**6 ))
print("Day ",find(f,24*60*60*10**6 ))
print("Year ",find(f,365*24*60*60*10**6 ))
print("Century ",find(f,100*365*24*60*60*10**6 ))
print("Mileniun ",find(f,1000*100*365*24*60*60*10**6 ))

Second  19
Minute   25
Hour  31
Day  36
Year  44
Century  51
Mileniun  61


$\bf{ Time\ Complexity\ n! }$

$Second$
$$  n!  \times 10^{-6} \ s\le 1\ s$$
$$  n! \le  10^{6} $$
$$  n \le  9 $$


$Minute$
$$  n! \times 10^{-6} \ s\le 60\ s$$
$$  n! \le 60 \times 10^{6} $$
$$  n \le  11 $$

$Hour$
$$  n! \times 10^{-6} \ s\le 3600\ s$$
$$  n! \le 3600 \times 10^{6}  = 3.6 \times 10^{10}$$
$$  n \le 12 $$


$Day$
$$  n! \times 10^{-6} \ s\le 84600\ s$$
$$  n! \le 84600 \times 10^{6}  = 8.46 \times 10^{11}$$
$$  n \le 13 $$

$Year$
$$  n! \times 10^{-6} \ s\le 31536000\ s$$
$$  n! \le 31536000 \times 10^{6}   = 3.153 \times 10^{14}$$
$$  n \le 16 $$

$Century$
$$  n! \times 10^{-6} \ s\le 3153600000\ s$$
$$  n! \le 3153600000 \times 10^{6}  = 3.153 \times 10^{16}$$
$$  n \le 17 $$

$Mileniun$
$$  n! \times 10^{-6} \ s\le 3153600000000\ s$$
$$  n! \le 3153600000000 \times 10^{6}  = 3.153 \times 10^{19}$$
$$  n \le 20 $$

In [16]:
import math
def f(n):
  return math.factorial(n)

def find(f,v):
  l = 1
  r = 100
  while True:
    m = int((l+r)/2)
    if ((f(m+1) > v) and (f(m) <= v) ) or (l == m) or (r== m):
      return int(m)
    elif f(m) < v:
      l = m
    else:
      r = m


print("Second ", find(f,10**6 ))
print("Minute  ",find(f,60*10**6 ))
print("Hour ",find(f,60*60*10**6 ))
print("Day ",find(f,24*60*60*10**6 ))
print("Year ",find(f,365*24*60*60*10**6 ))
print("Century ",find(f,100*365*24*60*60*10**6 ))
print("Mileniun ",find(f,1000*100*365*24*60*60*10**6 ))

Second  9
Minute   11
Hour  12
Day  13
Year  16
Century  17
Mileniun  20


$\bf{ Time\ Complexity\ n^n }$

$Second$
$$  n^n  \times 10^{-6} \ s\le 1\ s$$
$$  n^n \le  10^{6} $$
$$  n \le  7 $$

$Minute$
$$  n^n \times 10^{-6} \ s\le 60\ s$$
$$  n^n \le 60 \times 10^{6} $$
$$  n \le 8 $$

$Hour$
$$  n^n \times 10^{-6} \ s\le 3600\ s$$
$$  n^n \le 3600 \times 10^{6}  = 3.6 \times 10^{9}$$
$$  n \le 9 $$

$Day$
$$  n^n \times 10^{-6} \ s\le 84600\ s$$
$$  n^n \le 84600 \times 10^{6}  = 8.46 \times 10^{10}$$
$$  n \le 10 $$

$Year$
$$  n^n \times 10^{-6} \ s\le 31536000\ s$$
$$  n^n \le 31536000 \times 10^{6}   = 3.153 \times 10^{13}$$
$$  n \le 12 $$


$Century$
$$  n^n \times 10^{-6} \ s\le 3153600000\ s$$
$$  n^n \le 3153600000 \times 10^{6}  = 3.153 \times 10^{15}$$
$$  n \le 13 $$

$Mileniun$
$$  n^n \times 10^{-6} \ s\le 3153600000000\ s$$
$$  n^n \le 3153600000000 \times 10^{6}  = 3.153 \times 10^{18}$$
$$  n \le 15 $$


In [18]:
import math
def f(n):
  return n**n

def find(f,v):
  l = 1
  r = 100
  while True:
    m = int((l+r)/2)
    if ((f(m+1) > v) and (f(m) <= v) ) or (l == m) or (r== m):
      return int(m)
    elif f(m) < v:
      l = m
    else:
      r = m

print("Second =", find(f,10**6 ))
print("Minute  =",find(f,60*10**6 ))
print("Hour =",find(f,60*60*10**6 ))
print("Day =",find(f,24*60*60*10**6 ))
print("Year =",find(f,365*24*60*60*10**6 ))
print("Century =",find(f,100*365*24*60*60*10**6 ))
print("Mileniun =",find(f,1000*100*365*24*60*60*10**6 ))

Second = 7
Minute  = 8
Hour = 9
Day = 10
Year = 12
Century = 13
Mileniun = 15


# 2.(2.5  nanoseconds by instruction)

Asuuming that you can execute a 400 million instructions $400 \times 10^6$ every second, i.e. each instrcution takes 2.5  nanoseconds, 2.5 $\eta s  = 2.5 \times 10^{-9} s$. If you have algorithms take $\bf{ Time\ Complexity}$ steps to solve a problem of size $n$ as specified in column one. What is the size of the largest problem that you can solve in the time specified in row one?
$$
\begin{array}{|c|c|c|c|c|c|c|c|}
 \hline
 \bf{ Time\ Complexity} & Second  & Minute  & Hour & Day & Year & Century & Mileniun  \\
  \hline\\
\lg(n) &    &   &  &  &  &  &  \\
   \hline\\
\lg^2(n) &    &   &  &  &  &  &  \\
   \hline\\
\sqrt{n} &    &   &  &  &  &  &  \\
 \hline
 n &    &   &  &  &  &  &  \\
 \hline
 n\lg(n) &    &   &  &  &  &  &  \\
 \hline
 n^2 &    &   &  &  &  &  &  \\
 \hline
  n^3 &    &   &  &  &  &  &  \\
 \hline
  n^4 &    &   &  &  &  &  &  \\
 \hline
  2^n &    &   &  &  &  &  &  \\
 \hline
 n! &    &   &  &  &  &  &  \\
 \hline
 n^n &    &   &  &  &  &  &  \\
 \hline
\end{array}
$$

# 3. (1 noneosecond by instruction)

Asuuming that you can execute a 1000 million instructions $10^9$ every second, i.e. each instrcution takes 1  nanosecond, 1 $\eta s  = 10^{-9} s$. If you have algorithms take $\bf{ Time\ Complexity}$ steps to solve a problem of size $n$ as specified in column one. What is the size of the largest problem that you can solve in the time specified in row one?
$$
\begin{array}{|c|c|c|c|c|c|c|c|}
 \hline
 \bf{ Time\ Complexity} & Second  & Minute  & Hour & Day & Year & Century & Mileniun  \\
  \hline\\
\lg(n) &    &   &  &  &  &  &  \\
   \hline\\
\lg^2(n) &    &   &  &  &  &  &  \\
   \hline\\
\sqrt{n} &    &   &  &  &  &  &  \\
 \hline
 n &    &   &  &  &  &  &  \\
 \hline
 n\lg(n) &    &   &  &  &  &  &  \\
 \hline
 n^2 &    &   &  &  &  &  &  \\
 \hline
  n^3 &    &   &  &  &  &  &  \\
 \hline
  n^4 &    &   &  &  &  &  &  \\
 \hline
  2^n &    &   &  &  &  &  &  \\
 \hline
 n! &    &   &  &  &  &  &  \\
 \hline
 n^n &    &   &  &  &  &  &  \\
 \hline
\end{array}
$$